insitu_plugins/src/Heartbeat/Heartbeat.cpp

```diff
@@ -0,0 +1,109 @@
+#include <Heartbeat/Heartbeat.hpp>
+#include <Heartbeat/Heartbeat_dialog.hpp>
+
+using namespace heartbeat_filter;
+
+namespace temp_insitu_utils
+{
+// TODO: include from Label instead of copying
+void drawtorect(cv::Mat& mat, cv::Rect target, const std::string& str,
+                int face = cv::FONT_HERSHEY_PLAIN, int thickness = 1,
+                cv::Scalar color = cv::Scalar(255, 255, 255, 255))
+{
+    cv::Size rect = cv::getTextSize(str, face, 1.0, thickness, 0);
+    double scalex = (double)target.width / (double)rect.width;
+    double scaley = (double)target.height / (double)rect.height;
+    double scale = std::min(scalex, scaley);
+    int marginx =
+        scale == scalex ?
+            0 :
+            (int)((double)target.width * (scalex - scale) / scalex * 0.5);
+    int marginy =
+        scale == scaley ?
+            0 :
+            (int)((double)target.height * (scaley - scale) / scaley * 0.5);
+    cv::putText(
+        mat, str,
+        cv::Point(target.x + marginx, target.y + target.height - marginy), face,
+        scale, color, thickness, cv::LINE_AA, false);
+}
+}    // end namespace temp_insitu_utils
+
+namespace insitu_plugins
+{
+/*
+    Filter Implementation
+*/
+Heartbeat::Heartbeat(void)
+{
+    // TODO instantiation code
+}
+
+void Heartbeat::filterInit(void)
+{
+    settingsDialog = new HeartbeatDialog(this);
+    setSize(QSize(50, 50));
+
+    nh_ = getNodeHandle();
+    parser_ = RosIntrospection::Parser();
+
+    topic_name_ = getSettingsValue().get("topic", DEFAULT_TOPIC).asString();
+
+    topic_subscriber_ =
+        nh_.subscribe(topic_name_, 1, &Heartbeat::topicCB, this);
+}
+
+void Heartbeat::onDelete(void)
+{
+    // TODO cleanup code
+}
+
+const cv::Mat Heartbeat::apply(void)
+{
+    cv::Mat ret =
+        cv::Mat(height(), width(), CV_8UC4, cv::Scalar(255, 255, 255, 0));
+
+    double expected_hz =
+        getSettingsValue().get("rate", std::stod(DEFAULT_RATE)).asDouble();
+
+    cv::Scalar cvColor;
```

Comment:
Not required, but you could do something like:
```c++
int grad = ceil(255, (ros::Time::now() - last_msg_received_).toSec() * expected_hz * 255));
cvColor = cv::Scalar(grad, 255-grad, 0, 255); 
```
to create some sort of gradient for how old the heartbeat is.